types/webmidi/index.d.ts

```diff
@@ -4,193 +4,234 @@
+     * ID.
+     */
+    type MIDIInputMap = Map<string, MIDIInput>;
+
+    /**
+     * This is a maplike interface whose value is a MIDIOutput instance and key is its
+     * ID.
+     */
+    type MIDIOutputMap = Map<string, MIDIOutput>;
+
+    interface MIDIAccess extends EventTarget {
+        /**
+         * The MIDI input ports available to the system.
+         */
+        inputs: MIDIInputMap;
+
+        /**
+         * The MIDI output ports available to the system.
+         */
+        outputs: MIDIOutputMap;
+
+        /**
+         * The handler called when a new port is connected or an existing port changes the
+         * state attribute.
+         */
+        onstatechange(e: MIDIConnectionEvent): void;
+
+        addEventListener(
+            type: 'statechange',
+            listener: (this: this, e: MIDIConnectionEvent) => any,
+            options?: boolean | AddEventListenerOptions,
+        ): void;
+        addEventListener(
+            type: string,
+            listener: EventListenerOrEventListenerObject,
+            options?: boolean | AddEventListenerOptions,
+        ): void;
+
+        /**
+         * This attribute informs the user whether system exclusive support is enabled on
+         * this MIDIAccess.
+         */
+        sysexEnabled: boolean;
+    }
+
+    type MIDIPortType = 'input' | 'output';
+
+    type MIDIPortDeviceState = 'disconnected' | 'connected';
+
+    type MIDIPortConnectionState = 'open' | 'closed' | 'pending';
+
+    interface MIDIPort extends EventTarget {
+        /**
+         * A unique ID of the port. This can be used by developers to remember ports the
+         * user has chosen for their application.
+         */
+        id: string;
+
+        /**
+         * The manufacturer of the port.
+         */
+        manufacturer?: string;
+
+        /**
+         * The system name of the port.
+         */
+        name?: string;
+
+        /**
+         * A descriptor property to distinguish whether the port is an input or an output
+         * port.
+         */
+        type: MIDIPortType;
+
+        /**
+         * The version of the port.
+         */
+        version?: string;
+
+        /**
+         * The state of the device.
+         */
+        state: MIDIPortDeviceState;
+
+        /**
+         * The state of the connection to the device.
+         */
+        connection: MIDIPortConnectionState;
+
+        /**
+         * The handler called when an existing port changes its state or connection
+         * attributes.
+         */
+        onstatechange(e: MIDIConnectionEvent): void;
+
+        addEventListener(
+            type: 'statechange',
+            listener: (this: this, e: MIDIConnectionEvent) => any,
+            options?: boolean | AddEventListenerOptions,
+        ): void;
+        addEventListener(
+            type: string,
+            listener: EventListenerOrEventListenerObject,
+            options?: boolean | AddEventListenerOptions,
+        ): void;
```

Comment:
Is it necessary to define the non-discriminated overload here? Isn't it already part of `EventTarget`?

Reply:
Unfortunately, this seems to be required for the "implements EventTarget" to work. Otherwise, TS complains. I copied this approach from the EventTarget implementors in lib.dom.d.ts, so I think it's the best way to do this.